domains/client/block-preprocessor/src/inherents.rs

```diff
@@ -12,62 +12,153 @@
+        (),
+    >>::create_inherent_data_providers(
+        &create_inherent_data_providers, parent_hash, ()
+    )
+        .await?;
+    let mut inherent_data = InherentData::new();
+    inherent_data_providers
+        .provide_inherent_data(&mut inherent_data)
+        .await
+        .map_err(|err| {
+            sp_blockchain::Error::Application(Box::from(format!(
+                "failed to provide inherent data: {err:?}"
+            )))
+        })?;
+
+    Ok(inherent_data)
+}
+
+pub(crate) fn has_runtime_upgrade<CClient, CBlock, Block>(
+    consensus_client: &Arc<CClient>,
+    consensus_block_hash: CBlock::Hash,
+    domain_id: DomainId,
+) -> Result<bool, sp_blockchain::Error>
+where
+    CClient: ProvideRuntimeApi<CBlock> + HeaderBackend<CBlock>,
+    CClient::Api: DomainsApi<CBlock, NumberFor<Block>, Block::Hash>,
     CBlock: BlockT,
-    CClient: ProvideRuntimeApi<CBlock>,
+    Block: BlockT,
+{
+    let header = consensus_client.header(consensus_block_hash)?.ok_or(
+        sp_blockchain::Error::MissingHeader(format!(
+            "No header found for {consensus_block_hash:?}"
+        )),
+    )?;
+
+    let runtime_api = consensus_client.runtime_api();
+    let runtime_id = runtime_api
+        .runtime_id(consensus_block_hash, domain_id)?
+        .ok_or(sp_blockchain::Error::Application(Box::from(format!(
+            "No RuntimeId found for {domain_id:?}"
+        ))))?;
+
+    Ok(header
+        .digest()
+        .logs
+        .iter()
+        .filter_map(|log| log.as_domain_runtime_upgrade())
+        .any(|upgraded_runtime_id| upgraded_runtime_id == runtime_id))
+}
+
+fn maybe_runtime_upgrade<CClient, CBlock, Block>(
+    consensus_client: &Arc<CClient>,
+    consensus_block_hash: CBlock::Hash,
+    domain_id: DomainId,
+) -> Result<Option<Vec<u8>>, sp_blockchain::Error>
+where
+    CClient: ProvideRuntimeApi<CBlock> + HeaderBackend<CBlock>,
     CClient::Api: DomainsApi<CBlock, NumberFor<Block>, Block::Hash>,
-    DomainRuntimeApi: TimestampExtrinsicConstructor<Block>,
+    CBlock: BlockT,
+    Block: BlockT,
 {
-    let moment = consensus_client
-        .runtime_api()
-        .timestamp(consensus_block_hash)?;
+    let header = consensus_client.header(consensus_block_hash)?.ok_or(
+        sp_blockchain::Error::MissingHeader(format!(
+            "No header found for {consensus_block_hash:?}"
+        )),
+    )?;
+
+    let runtime_api = consensus_client.runtime_api();
+    let runtime_id = runtime_api
+        .runtime_id(consensus_block_hash, domain_id)?
+        .ok_or(sp_blockchain::Error::Application(Box::from(format!(
+            "No RuntimeId found for {domain_id:?}"
+        ))))?;
 
-    let inherent_exts =
-        vec![domain_runtime_api.construct_timestamp_extrinsic(domain_parent_hash, moment)?];
+    if header
+        .digest()
+        .logs
+        .iter()
+        .filter_map(|log| log.as_domain_runtime_upgrade())
+        .any(|upgraded_runtime_id| upgraded_runtime_id == runtime_id)
```

Comment:
This part seems to be duplicated with `has_runtime_upgrade`

Reply:
part of old change.